clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -557,11 +559,20 @@ caseCast _ e = return e
 
 -- | Push a cast over a Letrec into it's body
 letCast :: NormRewrite
-letCast _ (Cast (Letrec b) ty1 ty2) = do
+letCast _ c@(Cast (Letrec b) ty1 ty2) | shouldPushDown c = do
   let (binds,body) = unsafeUnbind b
   changed $ Letrec $ bind binds (Cast body ty1 ty2)
 letCast _ e = return e
 
+-- | Is this a 'Cast' that you be pushed down?
+shouldPushDown :: Term -> Bool
+shouldPushDown (Cast _ _ty1 ty2) = isIntegerTy ty2
+shouldPushDown _ = False
+
+argCastSpec :: NormRewrite
+argCastSpec ctx e@(App _ c@(Cast _ _ _)) | shouldPushDown c = specializeNorm ctx e
```

Comment:
Shouldn't we ensure that the cast is work-free?? so let-bind any work before we specialise.